signer/core/signed_data.go

```diff
@@ -0,0 +1,788 @@
+			item.Value = primitiveOutput
+		}
+		output = append(output, item)
+	}
+	return output
+}
+
+func formatPrimitiveValue(encType string, encValue interface{}) string {
+	switch encType {
+	case "address":
+		stringValue, _ := encValue.(string)
+		return common.HexToAddress(stringValue).String()
+	case "bool":
+		boolValue, _ := encValue.(bool)
+		return fmt.Sprintf("%t", boolValue)
+	case "bytes", "string":
+		return fmt.Sprintf("%s", encValue)
+	}
+	if strings.HasPrefix(encType, "bytes") {
+		return fmt.Sprintf("%s", encValue)
+	} else if strings.HasPrefix(encType, "uint") || strings.HasPrefix(encType, "int") {
+		bigIntValue, _ := encValue.(*big.Int)
+		return fmt.Sprintf("%d (0x%x)", bigIntValue, bigIntValue)
+	}
+	return "NA"
+}
+
+// NameValueType is a very simple struct with Name, Value and Type. It's meant for simple
+// json structures used to communicate signing-info about typed data with the UI
+type NameValueType struct {
+	Name  string      `json:"name"`
+	Value interface{} `json:"value"`
+	Typ   string      `json:"type"`
+}
+
+// Pprint returns a pretty-printed version of nvt
+func (nvt *NameValueType) Pprint(depth int) string {
+	output := bytes.Buffer{}
+	output.WriteString(strings.Repeat("\u00a0", depth*2))
+	output.WriteString(fmt.Sprintf("%s [%s]: ", nvt.Name, nvt.Typ))
+	if nvts, ok := nvt.Value.([]*NameValueType); ok {
+		output.WriteString("\n")
+		for _, next := range nvts {
+			sublevel := next.Pprint(depth + 1)
+			output.WriteString(sublevel)
+		}
+	} else {
+		output.WriteString(fmt.Sprintf("%s\n", nvt.Value))
+	}
+	return output.String()
+}
+
+// Validate checks if the types object is conformant to the specs
+func (t Types) Validate() error {
+	for typeKey, typeArr := range t {
+		for _, typeObj := range typeArr {
+			if typeKey == typeObj.Type {
+				return fmt.Errorf("type '%s' cannot reference itself", typeObj.Type)
+			}
+			if typeObj.isReferenceType() {
+				if _, exist := t[typeObj.Type]; !exist {
+					return fmt.Errorf("reference type '%s' is undefined", typeObj.Type)
+				}
+				if !typedDataReferenceTypeRegexp.MatchString(typeObj.Type) {
+					return fmt.Errorf("unknown reference type '%s", typeObj.Type)
+				}
+
+			} else if !typedDataRegexp.MatchString(typeObj.Type) {
+				return fmt.Errorf("unknown type '%s'", typeObj.Type)
+			}
+		}
+	}
+	return nil
+}
+
+// Validate checks if the given domain is valid, i.e. contains at least
+// the minimum viable keys and values
+func (domain *TypedDataDomain) Validate() error {
+	if domain.ChainId == big.NewInt(0) {
+		return errors.New("chainId must be specified according to EIP-155")
+	}
+
+	if len(domain.Name) == 0 && len(domain.Version) == 0 && len(domain.VerifyingContract) == 0 && len(domain.Salt) == 0 {
+		return errors.New("domain is undefined")
+	}
+
+	return nil
+}
+
+// Map is a helper function to generate a map version of the domain
+func (domain *TypedDataDomain) Map() map[string]interface{} {
```

Comment:
The spec says that domain descriptions could be extended in the future. I would prefer something that copies (if necessary) the data from the map without making too many assumptions about its content.